### ISSUE TITLE
vmap of pmap gives wrong shape

### ISSUE THREAD

**Issue by JasperSnoek** [google/jax#2787]
When I do a batched pmap (vmap over pmap), I can access the batch dimension inside the vmap (and it hides another dim of the array).  See repro below:

```
from jax import lax, pmap, vmap, map
def vv(x, y):
  """Vector-vector multiply"""
  return np.vdot(x, y)

def distributed_matrix_vector(x, y, distributed=True):
  """Matrix vector multiply. First batch it and then row by row"""
  fv = lambda z: lax.map(lambda j: vv(j, y), z)
  if distributed:
    res = pmap(fv)(x.reshape((jax.device_count(), -1) + tuple(x.shape[1:])))
  else:
    res = fv(x)
  print("Inside matrix vector code %s (should be %s)" % (res.shape, (x.shape[0], 1)))
  return res

print("Not vmapped")
key = random.PRNGKey(1)
x = random.normal(key, (800, 50))
y = random.normal(key, (50, 1))
result = distributed_matrix_vector(x, y)
print("Outside matrix vector code", result.shape)

print("Vmapped")
batched_mvm = vmap(lambda b: distributed_matrix_vector(x, b, True), in_axes=(0,))
y = random.normal(key, (10, 50))
result = batched_mvm(y)
print("Outside matrix vector code", result.shape)

# Not vmapped
# Inside matrix vector code (8, 100) (should be (800, 1))
# Outside matrix vector code (8, 100)
# Vmapped
# Inside matrix vector code (8, 10) (should be (800, 1))
# Outside matrix vector code (100, 8, 10)```

**Comment by mattjj** [comment]
I think perhaps we want to reshape `res` before returning it, to merge the pmapped axis back in. Also, if we use `np.dot` instead of `np.vdot` (because the latter when called on a `(50,)` input and a `(50, 1)` input will return a scalar result, flattening the second argument) we get nicer shapes:

```python
def vv(x, y):
  """Vector-vector multiply"""
  return np.dot(x, y)

def distributed_matrix_vector(x, y):
  """Matrix vector multiply. First batch it and then row by row"""
  fv = lambda z: lax.map(lambda j: vv(j, y), z)  # or vmap
  res = pmap(fv)(x.reshape((jax.device_count(), -1) + tuple(x.shape[1:])))
  res = res.reshape(res.shape[0] * res.shape[1], *res.shape[2:])
  print("Inside matrix vector code %s (should be %s)" % (res.shape, (x.shape[0], 1)))
  return res
```

Then the first "Not vmapped" section prints like this:

```
Not vmapped
Inside matrix vector code (800, 1) (should be (800, 1))
Outside matrix vector code (800, 1)
```

Will look at the second bit now.

**Comment by mattjj** [comment]
Second bit is definitely a JAX bug! One that I'm surprised lasted this long, but I have a vague feeling I knew this wasn't covered by our tests. Thanks for finding our bugs, Jaxpr Snoek.

The issue was basically that we reused `BatchTrace.post_process_call`, a method for handling some details for `jit`, for `pmap` as well. But in some cases we need to handle `pmap` differently. We need a `post_process_map`!

I'll take this opportunity to refactor some stuff...
